File: bingo/expressions/agraph/simplification/interpreter.py
"""Translate between command-array stacks and CAS expression trees.

Adapted for the ``expressions`` package where CONSTANT and INTEGER
values are stored in external tuples (not embedded in the command
array).
"""

import numpy as np

from ..operators import IS_TERMINAL_MAP, IS_ARITY_2_MAP, CONSTANT, INTEGER, VARIABLE
from .cas_expression import CASExpression, get_interned_integer, get_interned_variable


def build_cas_expression(stack, constants, integers):
    """Translate a command array into a CAS expression tree.

    Memoizes by row index so that shared sub-expressions in the DAG
    are built only once.

    Parameters
    ----------
    stack : Nx3 numpy array of int
        The command array encoding the equation.
    constants : tuple of float
        Constant values indexed by CONSTANT node params.
    integers : tuple of int
        Integer values indexed by INTEGER node params.

    Returns
    -------
    CASExpression
    """
    memo = {}
    return _build_expression_recursive(stack, constants, integers, len(stack) - 1, memo)


def _build_expression_recursive(stack, constants, integers, location, memo):
    if location in memo:
        return memo[location]

    operator = int(stack[location, 0])
    param_1 = int(stack[location, 1])
    param_2 = int(stack[location, 2])

    if IS_TERMINAL_MAP[operator]:
        if operator == CONSTANT:
            # Store an opaque constant index — constant folding uses
            # this to track which constants can be merged.
            result = CASExpression(CONSTANT, [param_1])
        elif operator == INTEGER:
            # Store the actual integer *value* for integer arithmetic
            # during automatic simplification.
            value = integers[param_1] if param_1 < len(integers) else 0
            result = get_interned_integer(value)
        else:
            # VARIABLE — store the column index
            result = get_interned_variable(param_1)
        memo[location] = result
        return result

    operands = [_build_expression_recursive(stack, constants, integers, param_1, memo)]
    if IS_ARITY_2_MAP[operator]:
        operands.append(
            _build_expression_recursive(stack, constants, integers, param_2, memo)
        )
    result = CASExpression(operator, operands)
    memo[location] = result
    return result


def build_simplified_cas_expression(stack, constants, integers):
    """Build a CAS tree and simplify each node as it is constructed.

    This fuses the bottom-up tree build with
    :func:`automatic_simplify`, eliminating a separate recursive
    simplification pass.  The memo stores *already-simplified* nodes so
    shared sub-expressions are only simplified once.

    Parameters
    ----------
    stack : Nx3 numpy array of int
        The command array encoding the equation.
    constants : tuple of float
        Constant values indexed by CONSTANT node params.
    integers : tuple of int
        Integer values indexed by INTEGER node params.

    Returns
    -------
    CASExpression
    """
    from .automatic_simplification import SIMPLIFICATION_FUNCTIONS

    memo = {}
    return _build_simplified_recursive(
        stack,
        constants,
        integers,
        len(stack) - 1,
        memo,
        SIMPLIFICATION_FUNCTIONS,
    )


def _build_simplified_recursive(stack, constants, integers, location, memo, simp_funcs):
    if location in memo:
        return memo[location]

    operator = int(stack[location, 0])
    param_1 = int(stack[location, 1])
    param_2 = int(stack[location, 2])

    if IS_TERMINAL_MAP[operator]:
        if operator == CONSTANT:
            result = CASExpression(CONSTANT, [param_1])
        elif operator == INTEGER:
            value = integers[param_1] if param_1 < len(integers) else 0
            result = get_interned_integer(value)
        else:
            result = get_interned_variable(param_1)
        memo[location] = result
        return result

    # Operands are already simplified (bottom-up + memo).
    operands = [
        _build_simplified_recursive(
            stack, constants, integers, param_1, memo, simp_funcs
        )
    ]
    if IS_ARITY_2_MAP[operator]:
        operands.append(
            _build_simplified_recursive(
                stack, constants, integers, param_2, memo, simp_funcs
            )
        )
    node = CASExpression(operator, operands)
    # Dispatch to the operator-specific simplifier.
    result = simp_funcs[operator](node)
    memo[location] = result
    return result


def build_agraph_stack(expression, original_constants):
    """Translate a CAS expression tree back into a command array.

    Parameters
    ----------
    expression : CASExpression
        The simplified CAS expression.
    original_constants : tuple of float
        The original constant values.  Each CAS ``CONSTANT`` node
        references an index into this tuple; unknown indices default
        to ``1.0``.

    Returns
    -------
    tuple
        ``(command_array, constants, integers, const_idx_map)`` — a
        uint8 command array, the corresponding constant/integer tuples,
        and a dict mapping each original CAS constant index to the new
        sequential index in the output constants tuple.
    """
    stack_dict = {}
    const_list = []
    int_list = []
    const_idx_map = {}  # old_idx -> new_idx (dedup same original index)
    int_val_map = {}  # value -> new_idx (dedup same integer value)
    _build_stack_recursive(
        expression,
        stack_dict,
        original_constants,
        const_list,
        int_list,
        const_idx_map,
        int_val_map,
    )

    stack = np.empty((len(stack_dict), 3), dtype=np.uint8)
    for command, loc in stack_dict.items():
        stack[loc] = command
    return stack, tuple(const_list), tuple(int_list), const_idx_map


def _build_stack_recursive(
    expression,
    stack_dict,
    original_constants,
    const_list,
    int_list,
    const_idx_map,
    int_val_map,
):
    operator = expression.operator

    if operator == CONSTANT:
        old_idx = expression.operands[0]
        if old_idx in const_idx_map:
            new_idx = const_idx_map[old_idx]
        else:
            new_idx = len(const_list)
            if old_idx < len(original_constants):
                value = float(original_constants[old_idx])
            else:
                value = 1.0
            const_list.append(value)
            const_idx_map[old_idx] = new_idx
        command = (operator, new_idx, new_idx)
        return _add_command_to_stack_dict(command, stack_dict)

    if operator == INTEGER:
        value = expression.operands[0]
        if value in int_val_map:
            new_idx = int_val_map[value]
        else:
            new_idx = len(int_list)
            int_list.append(value)
            int_val_map[value] = new_idx
        command = (operator, new_idx, new_idx)
        return _add_command_to_stack_dict(command, stack_dict)

    if operator == VARIABLE:
        col = expression.operands[0]
        command = (operator, col, col)
        return _add_command_to_stack_dict(command, stack_dict)

    # Non-terminal
    operand_locations = [
        _build_stack_recursive(
            operand,
            stack_dict,
            original_constants,
            const_list,
            int_list,
            const_idx_map,
            int_val_map,
        )
        for operand in expression.operands
    ]

    if len(operand_locations) == 1:
        command = (operator, operand_locations[0], operand_locations[0])
        return _add_command_to_stack_dict(command, stack_dict)

    if len(operand_locations) == 2:
        command = (operator, operand_locations[0], operand_locations[1])
        return _add_command_to_stack_dict(command, stack_dict)

    # Associative operators with >2 operands (flattened sums/products)
    if not expression.is_constant_valued and expression.operands[0].is_constant_valued:
        loc = _add_associative_operators_to_stack(
            operator, operand_locations[1:], stack_dict
        )
        command = (operator, operand_locations[0], loc)
        return _add_command_to_stack_dict(command, stack_dict)

    return _add_associative_operators_to_stack(operator, operand_locations, stack_dict)


def _add_command_to_stack_dict(command, stack_dict):
    if command in stack_dict:
        return stack_dict[command]
    loc = len(stack_dict)
    stack_dict[command] = loc
    return loc


def _add_associative_operators_to_stack(operator, operand_locs, stack_dict):
    if len(operand_locs) == 1:
        return operand_locs[0]
    operand_div = len(operand_locs) // 2
    loc_1 = _add_associative_operators_to_stack(
        operator, operand_locs[:operand_div], stack_dict
    )
    loc_2 = _add_associative_operators_to_stack(
        operator, operand_locs[operand_div:], stack_dict
    )
    command = (operator, loc_1, loc_2)
    return _add_command_to_stack_dict(command, stack_dict)

File: bingo/expressions/agraph/cppagraph/src/interpreter.cpp
/**
 * @file interpreter.cpp
 * @brief Stack ↔ CAS tree translation — port of interpreter.py.
 */

#include "cppagraph/interpreter.h"
#include "cppagraph/automatic_simplification.h"

#include <map>
#include <set>
#include <tuple>
#include <unordered_map>

namespace cppagraph {

static inline uint8_t u8(Op o) { return static_cast<uint8_t>(o); }

// ================================================================== //
//  build_cas_expression                                               //
// ================================================================== //

static CASExprPtr _build_expression_recursive(
    const StackMatrix& stack,
    const std::vector<double>& constants,
    const std::vector<int>& integers,
    int location,
    std::unordered_map<int, CASExprPtr>& memo,
    std::set<int>& visiting)
{
    auto it = memo.find(location);
    if (it != memo.end()) return it->second;

    // Cycle guard.
    if (visiting.count(location)) {
        auto result = interned_variable(0);
        memo[location] = result;
        return result;
    }
    visiting.insert(location);

    uint8_t op     = stack(location, 0);
    int     param1 = stack(location, 1);
    int     param2 = stack(location, 2);

    CASExprPtr result;

    if (IS_TERMINAL[op]) {
        if (op == u8(Op::CONSTANT)) {
            result = std::make_shared<CASExpression>(u8(Op::CONSTANT), param1);
        } else if (op == u8(Op::INTEGER)) {
            int value = (param1 < static_cast<int>(integers.size()))
                        ? integers[param1] : 0;
            result = interned_integer(value);
        } else {
            result = interned_variable(param1);
        }
        memo[location] = result;
        visiting.erase(location);
        return result;
    }

    std::vector<CASExprPtr> operands;
    operands.push_back(
        _build_expression_recursive(stack, constants, integers, param1, memo, visiting));
    if (IS_ARITY_2[op]) {
        operands.push_back(
            _build_expression_recursive(stack, constants, integers, param2, memo, visiting));
    }

    // Normalise SQUARE/CUBE into POWER.
    if (op == u8(Op::SQUARE)) {
        result = std::make_shared<CASExpression>(
            u8(Op::POWER),
            std::vector<CASExprPtr>{operands[0], interned_integer(2)});
    } else if (op == u8(Op::CUBE)) {
        result = std::make_shared<CASExpression>(
            u8(Op::POWER),
            std::vector<CASExprPtr>{operands[0], interned_integer(3)});
    } else {
        result = std::make_shared<CASExpression>(op, std::move(operands));
    }

    memo[location] = result;
    visiting.erase(location);
    return result;
}

CASExprPtr build_cas_expression(
    const StackMatrix& stack,
    const std::vector<double>& constants,
    const std::vector<int>& integers)
{
    std::unordered_map<int, CASExprPtr> memo;
    std::set<int> visiting;
    return _build_expression_recursive(
        stack, constants, integers, static_cast<int>(stack.rows()) - 1, memo, visiting);
}

// ================================================================== //
//  build_simplified_cas_expression                                    //
// ================================================================== //

static CASExprPtr _build_simplified_recursive(
    const StackMatrix& stack,
    const std::vector<double>& constants,
    const std::vector<int>& integers,
    int location,
    std::unordered_map<int, CASExprPtr>& memo,
    const std::unordered_map<uint8_t, SimplifyFn>& simp_funcs,
    std::set<int>& visiting)
{
    auto it = memo.find(location);
    if (it != memo.end()) return it->second;

    if (visiting.count(location)) {
        auto result = interned_variable(0);
        memo[location] = result;
        return result;
    }
    visiting.insert(location);

    uint8_t op     = stack(location, 0);
    int     param1 = stack(location, 1);
    int     param2 = stack(location, 2);

    CASExprPtr result;

    if (IS_TERMINAL[op]) {
        if (op == u8(Op::CONSTANT)) {
            result = std::make_shared<CASExpression>(u8(Op::CONSTANT), param1);
        } else if (op == u8(Op::INTEGER)) {
            int value = (param1 < static_cast<int>(integers.size()))
                        ? integers[param1] : 0;
            result = interned_integer(value);
        } else {
            result = interned_variable(param1);
        }
        memo[location] = result;
        visiting.erase(location);
        return result;
    }

    // Operands are already simplified (bottom-up + memo).
    std::vector<CASExprPtr> operands;
    operands.push_back(
        _build_simplified_recursive(
            stack, constants, integers, param1, memo, simp_funcs, visiting));
    if (IS_ARITY_2[op]) {
        operands.push_back(
            _build_simplified_recursive(
                stack, constants, integers, param2, memo, simp_funcs, visiting));
    }

    // Normalise SQUARE/CUBE into POWER, then simplify.
    uint8_t effective_op = op;
    if (op == u8(Op::SQUARE)) {
        auto node = std::make_shared<CASExpression>(
            u8(Op::POWER),
            std::vector<CASExprPtr>{operands[0], interned_integer(2)});
        effective_op = u8(Op::POWER);
        auto sit = simp_funcs.find(effective_op);
        result = (sit != simp_funcs.end()) ? sit->second(node) : node;
    } else if (op == u8(Op::CUBE)) {
        auto node = std::make_shared<CASExpression>(
            u8(Op::POWER),
            std::vector<CASExprPtr>{operands[0], interned_integer(3)});
        effective_op = u8(Op::POWER);
        auto sit = simp_funcs.find(effective_op);
        result = (sit != simp_funcs.end()) ? sit->second(node) : node;
    } else {
        auto node = std::make_shared<CASExpression>(op, std::move(operands));
        auto sit = simp_funcs.find(op);
        result = (sit != simp_funcs.end()) ? sit->second(node) : node;
    }

    memo[location] = result;
    visiting.erase(location);
    return result;
}

CASExprPtr build_simplified_cas_expression(
    const StackMatrix& stack,
    const std::vector<double>& constants,
    const std::vector<int>& integers)
{
    auto& simp_funcs = simplification_functions();
    std::unordered_map<int, CASExprPtr> memo;
    std::set<int> visiting;
    return _build_simplified_recursive(
        stack, constants, integers, static_cast<int>(stack.rows()) - 1,
        memo, simp_funcs, visiting);
}

// ================================================================== //
//  build_agraph_stack                                                 //
// ================================================================== //

using CommandKey = std::tuple<uint8_t, int, int>;

static int _add_command_to_stack_dict(
    const CommandKey& command,
    std::map<CommandKey, int>& stack_dict)
{
    auto it = stack_dict.find(command);
    if (it != stack_dict.end()) return it->second;
    int loc = static_cast<int>(stack_dict.size());
    stack_dict[command] = loc;
    return loc;
}

static int _add_associative_operators_to_stack(
    uint8_t op,
    const std::vector<int>& operand_locs,
    std::map<CommandKey, int>& stack_dict)
{
    if (operand_locs.size() == 1) return operand_locs[0];
    size_t div = operand_locs.size() / 2;
    std::vector<int> left(operand_locs.begin(), operand_locs.begin() + div);
    std::vector<int> right(operand_locs.begin() + div, operand_locs.end());
    int loc1 = _add_associative_operators_to_stack(op, left, stack_dict);
    int loc2 = _add_associative_operators_to_stack(op, right, stack_dict);
    return _add_command_to_stack_dict({op, loc1, loc2}, stack_dict);
}

static int _build_stack_recursive(
    const CASExprPtr& expression,
    std::map<CommandKey, int>& stack_dict,
    const std::vector<double>& original_constants,
    std::vector<double>& const_list,
    std::vector<int>& int_list,
    std::map<int, int>& const_idx_map,
    std::map<int, int>& int_val_map)
{
    uint8_t op = expression->op();

    if (op == u8(Op::CONSTANT)) {
        int old_idx = expression->terminal_param();
        auto it = const_idx_map.find(old_idx);
        int new_idx;
        if (it != const_idx_map.end()) {
            new_idx = it->second;
        } else {
            new_idx = static_cast<int>(const_list.size());
            double value = (old_idx < static_cast<int>(original_constants.size()))
                           ? original_constants[old_idx] : 1.0;
            const_list.push_back(value);
            const_idx_map[old_idx] = new_idx;
        }
        return _add_command_to_stack_dict({op, new_idx, new_idx}, stack_dict);
    }

    if (op == u8(Op::INTEGER)) {
        int value = expression->terminal_param();
        auto it = int_val_map.find(value);
        int new_idx;
        if (it != int_val_map.end()) {
            new_idx = it->second;
        } else {
            new_idx = static_cast<int>(int_list.size());
            int_list.push_back(value);
            int_val_map[value] = new_idx;
        }
        return _add_command_to_stack_dict({op, new_idx, new_idx}, stack_dict);
    }

    if (op == u8(Op::VARIABLE)) {
        int col = expression->terminal_param();
        return _add_command_to_stack_dict({op, col, col}, stack_dict);
    }

    // Non-terminal.
    auto& operands = expression->operands();
    std::vector<int> operand_locations;
    operand_locations.reserve(operands.size());
    for (auto& child : operands) {
        operand_locations.push_back(
            _build_stack_recursive(
                child, stack_dict, original_constants,
                const_list, int_list, const_idx_map, int_val_map));
    }

    if (operand_locations.size() == 1) {
        return _add_command_to_stack_dict(
            {op, operand_locations[0], operand_locations[0]}, stack_dict);
    }

    if (operand_locations.size() == 2) {
        return _add_command_to_stack_dict(
            {op, operand_locations[0], operand_locations[1]}, stack_dict);
    }

    // Associative operators with >2 operands.
    if (!expression->is_constant_valued()
        && operands[0]->is_constant_valued()) {
        std::vector<int> rest(operand_locations.begin() + 1, operand_locations.end());
        int loc = _add_associative_operators_to_stack(op, rest, stack_dict);
        return _add_command_to_stack_dict({op, operand_locations[0], loc}, stack_dict);
    }

    return _add_associative_operators_to_stack(op, operand_locations, stack_dict);
}

BuildStackResult build_agraph_stack(
    const CASExprPtr& expression,
    const std::vector<double>& original_constants)
{
    std::map<CommandKey, int> stack_dict;
    std::vector<double> const_list;
    std::vector<int> int_list;
    std::map<int, int> const_idx_map;
    std::map<int, int> int_val_map;

    _build_stack_recursive(
        expression, stack_dict, original_constants,
        const_list, int_list, const_idx_map, int_val_map);

    int n = static_cast<int>(stack_dict.size());
    StackMatrix stack(n, 3);
    for (auto& [cmd, loc] : stack_dict) {
        stack(loc, 0) = std::get<0>(cmd);
        stack(loc, 1) = static_cast<uint8_t>(std::get<1>(cmd));
        stack(loc, 2) = static_cast<uint8_t>(std::get<2>(cmd));
    }

    return {std::move(stack), std::move(const_list),
            std::move(int_list), std::move(const_idx_map)};
}

}  // namespace cppagraph
